{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c3febc-10e6-484a-ac57-f07ee1eaf74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "np.set_printoptions(legacy = '1.25')\n",
    "\n",
    "import diem\n",
    "import time\n",
    "\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffedcbcd-8426-41a5-85bf-44096c2a758d",
   "metadata": {},
   "source": [
    "# Prepare Data for `diem` analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e991155a-ecd2-428c-9898-1c5669d6f394",
   "metadata": {},
   "source": [
    "## Convert vcf file to diem input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bafcd0c8-8010-4fcc-a244-2ab8bd54d100",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfFile = './data/example_dataset.vcf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db35e651-363c-419a-92e7-9fae92271448",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('vcf2diem '+vcfFile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc8c5069-c465-4186-9c5e-e4bf37fd171f",
   "metadata": {},
   "source": [
    "## Update Ploidy information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da3fadd3-e7b3-4118-ad1e-30b164415f59",
   "metadata": {},
   "source": [
    "vcf2diem generates a meta data file that provides information on chromosome lengths as well as the ploidy of each chromosome for each individual. \n",
    "\n",
    "By default, it assigns a ploidy of 2 across all individuals and chromosomes.  However, for some chromosomes, this will not be the case.  For example, in an XY system, the male is haploid for the X chromosome.  \n",
    "\n",
    "In our example, we provide a file to update and correct this information.\n",
    "\n",
    "We apply this update using the update_ploidy function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "620a00a3-d869-427f-ab90-5f6732aaef8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "metaRawFile = './data/example_dataset.vcf.gz.diem_meta.bed'\n",
    "\n",
    "ploidyFile = './user_files/chrom_ploidy.tsv'\n",
    "\n",
    "metaCorrected = './data/example_dataset.vcf.gz.diem_meta_corrected.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c219757f-1ecc-41df-be77-88c9a77cb19b",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.update_ploidy(ploidyFile,metaFilePathIn=metaRawFile,metaFilePathOut=metaCorrected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21c15d1c-a488-4391-83b5-a413eb7401fb",
   "metadata": {},
   "source": [
    "# Reading data into diemPy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0f93caa-05f1-499b-a4c3-f1787fe33872",
   "metadata": {},
   "source": [
    "We now have our inbut bed file and our *corrected* meta data file.  \n",
    "\n",
    "We can read that in as a 'raw' or 'unpolarized' DiemType object.\n",
    "\n",
    "The DiemType is a class which holds the data and provides the core functionality of diem. Refer to documentation for full details.\n",
    "\n",
    "Here, we will simply show how it is used in a standard workflow. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a0a17d-a960-4cbd-8cba-3b3bde606536",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedInFile = './data/example_dataset.vcf.gz.diem_input.bed'\n",
    "metaCorrected = './data/example_dataset.vcf.gz.diem_meta_corrected.bed'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "763b2142-1a54-428b-828a-ae0466270a17",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dRaw = diem.read_diem_bed(bedInFile,metaCorrected)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1585b19-08cf-4bac-8ec3-fa8731f4554f",
   "metadata": {},
   "source": [
    "Let's have a look at the variables (':ivar') that store the information for diem:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d38460-56d7-4e2d-af95-6ea79344e92b",
   "metadata": {},
   "outputs": [],
   "source": [
    "?dRaw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28eb6736-fd8f-4f43-a770-7f1905667c84",
   "metadata": {},
   "source": [
    "# Polarization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "939cc568-8117-4a91-9157-cbbd6c86647b",
   "metadata": {},
   "source": [
    "To polarize the data, we simply call the polarize() function of the DiemType object we created above.\n",
    "\n",
    "Note:\n",
    "- that it defaults to a parallel evaluation using the full number of available cores. We can change the number of cores manually using the 'ncores=' optional argument.  If n=1, it performs a serial polarization of the data.\n",
    "- it returns a separate *copy* of the DiemType object that is now polarized. That is, the raw object is not changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85529778-951e-4cd1-96f8-11e752a5fb16",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPol = dRaw.polarize(ncores=4,epsilon = 0.999999999999) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "022cf82a-4e69-48b2-88c6-60cfb91a3363",
   "metadata": {},
   "source": [
    "We can now save this to a bed file which has the same format as the input file but added information from polarization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793c41fc-f05c-4a73-91b3-3eed37ad9c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedOutFile = './output/diem_output.bed'\n",
    "diem.write_polarized_bed(bedInFile,bedOutFile,dPol)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27c2421a-0f11-4a05-80d4-ec8dfae47e4e",
   "metadata": {},
   "source": [
    "We can also save this directly as a diemtype object. \n",
    "\n",
    "It uses less memory, and it is how we will work with the data moving forward.\n",
    "\n",
    "However, it is only useable within the diem environment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5214cdd0-0a9f-495b-ac90-471bf8a08f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "polarizedDiemTypeOutFile = './output/example_polarized.diemtype'\n",
    "\n",
    "diem.save_DiemType(dPol,polarizedDiemTypeOutFile)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6259607d-3276-4085-833e-fa3ce77c00f0",
   "metadata": {},
   "source": [
    "If we have already polarized the data, we can load the data back in either way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19a9750-d95e-428b-90ec-1b1c6ec6a012",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPol = diem.load_DiemType(polarizedDiemTypeOutFile) \n",
    "#or\n",
    "dPol = diem.read_diem_bed(bedOutFile,metaCorrected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e17d1432-aebd-4b0e-bfe9-e4997d84e59f",
   "metadata": {},
   "source": [
    "# Hybrid Indices and Sorting "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f0c6dc-5eff-4372-9b3d-aba70ad25b70",
   "metadata": {},
   "source": [
    "the `diem` algorithm sorts individuals by hybrid index and seeks to maximize the differences between two sets of individuals on either side of a barrier to gene flow.\n",
    "\n",
    "although the hybrid indices are calculated on polarization, the individuals are not sorted by hybrid index as we can see here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee65f0a-5342-4f70-ba35-bdc04743b5bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,4))\n",
    "plt.plot(dPol.HIs,marker='.',)\n",
    "plt.ylim(0,1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bde02d9-2e3c-4df3-b26e-ae3e97787d36",
   "metadata": {},
   "source": [
    "We can use the sort() function to sort the individuals. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02504da1-756a-4d70-8bba-209ae884da43",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPol.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a3b4468-493e-4955-b7fc-e2f7ddf8ed59",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,4))\n",
    "plt.plot(dPol.HIs,marker='.',)\n",
    "plt.ylim(0,1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1abbaf54-4d9d-4c6d-acb7-4db75a1236a9",
   "metadata": {},
   "source": [
    "Now that we have the "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ca15fd-0507-40e0-adb5-85468a0127c3",
   "metadata": {},
   "source": [
    "# Plotting a Painting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b98c1691-652e-40a6-8476-0a80fb6cfa46",
   "metadata": {},
   "source": [
    "Now that we have polarized and sorted the data, let's have a look at it using the plot_painting() function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c48704-3486-48fc-9b1f-1e0b73102586",
   "metadata": {},
   "outputs": [],
   "source": [
    "#here we are plotting chromosome 9\n",
    "diem.plot_painting(dPol.DMBC[9])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a86f3ffd-b727-413e-be1a-145d625462c4",
   "metadata": {},
   "source": [
    "Simply for illustration, let's compare this to the null polarization that diem started with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e544bf64-4446-4244-830d-f131ff2cbcd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(diem.diemtype.flip_polarity(dPol.DMBC[9],dPol.initialPolByChr[9]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c108d8d1-6d60-49e5-a654-65f0a1fc2e99",
   "metadata": {},
   "source": [
    "# The Diagnostic Index (DI) and Thresholding"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "939cdfab-a788-414d-8c37-a56dd14897d2",
   "metadata": {},
   "source": [
    "## The DI distribution "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fb55c21-bc1f-44fc-b769-9410a810009b",
   "metadata": {},
   "source": [
    "Here we plot the distribution of diagnostic indices across the data.  For each of the final marker configurations, the diagnostic index tells use how well the marker matches does at explaining our data.  The least-negative value is the best.\n",
    "\n",
    "In teal, we plot a histogram of the DI values for each unique marker.\n",
    "\n",
    "In magenta, the counts of each marker are taken into account."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bef5ff0-3eab-48b3-9787-f992113c89fd",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "DIValsWithRepeats = np.hstack(dPol.DIByChr)\n",
    "DIValsUnique = np.unique(DIValsWithRepeats)\n",
    "plt.figure(figsize=(5,4))\n",
    "(counts,bins,patches) = plt.hist(DIValsWithRepeats,bins=500,color='darkmagenta')\n",
    "plt.hist(DIValsUnique,bins=bins,color='lightseagreen')\n",
    "plt.yscale('log')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9ecce6e-f619-4fa7-b289-9f41cbd2527f",
   "metadata": {},
   "source": [
    "We provide a function that will characterize the marker configurations, and provide information about each:  the count, the DI, the number of differences from the 'ideal marker' (the one with the highest DI) and how  many of the individuals have missing data for that marker. It returns a pandas dataframe.\n",
    "\n",
    "This could be useful for exploring the data in greater detail when deciding on a threshold value in later steps.\n",
    "\n",
    "**note** that this could be memory and time intensive to generate if there are many marker types.  We recommend characterizing only a subset of the most-informative markers to be on the safe side.  It is those markers that we will retain through thresholding later.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00fbb496-1607-4ec1-8d9f-69ebb056a966",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfMarkers = diem.characterize_markers(dPol.apply_threshold(-70))\n",
    "dfMarkers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87a19d32-8533-4aba-ab43-b2d3c14bbcfb",
   "metadata": {},
   "source": [
    "## applying the final threshold chosen"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3842f21-6b2a-45d8-9067-4ff29f0bfa92",
   "metadata": {},
   "source": [
    "The DI distribution for every dataset will be unique. However, many of the markers in the data are SNPs that are not informative about the barrier.\n",
    "\n",
    "For downstream analysis, we want to choose a threshold for retaining markers.  \n",
    "\n",
    "Here, we pick a threshold of -70.  Thresholding returns a copy of the data, and the propoertion of markers retained for each chromsoome is reported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58625c4c-0173-469f-9cb1-b25b5b076633",
   "metadata": {},
   "outputs": [],
   "source": [
    "dThresh = dPol.apply_threshold(-60)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "287dd38e-8b74-45b2-b28d-3e837b6c005e",
   "metadata": {},
   "source": [
    "After thresholding, the hybrid indices may change.  We can update these and re-sort the individuals in the data using the sort() function again:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a9bc3c-196d-4f25-a0ae-71b271252c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "dThresh.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157827df-3d7e-4eb5-9551-2189e1b97ee4",
   "metadata": {},
   "source": [
    "Let's compare the before (b) and after (r) thresholding:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c0eaca-82a3-406e-be47-97b31357be32",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,4))\n",
    "plt.plot(dPol.HIs,marker='.',color='b')\n",
    "plt.plot(dThresh.HIs,marker='.',color='r')\n",
    "plt.ylim(0,1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "528b0f20-f9b3-415e-958c-e2b48d750d2b",
   "metadata": {},
   "source": [
    "After thresholding, there will be fewer markers present. \n",
    "Let's plot the painting after thresholding and see how much cleaner it looks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c98a7da5-7994-4cf7-9be5-d0c0e9a29f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(dThresh.DMBC[9])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f20ba98-4393-490b-a92a-67667288ebb1",
   "metadata": {},
   "source": [
    "# Smoothing the Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e285eea4-8574-4d19-9f54-e1a11da7360c",
   "metadata": {},
   "source": [
    "After thresholding, we will see much clearer tracts representing shared ancestry with respect to the barrier.  However, there will be some bits of missing data, mis-called sites, etc. that will break up the tracts. We want to smooth this over for later analyses. We also care about tracts generated by cross-over recombination, but gene conversion events could also cause short changes in state and break up our tracts.  Therefore, we want to smooth over this noise and this signal in the data.\n",
    "\n",
    "To pick a smoothing scale, we examine how the data changes for different values.  Below, we calculate the number of sites that change across the whole dataset using the laplacian smoothign kernel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f09ff31e-aed0-4871-939b-910d3c530b91",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "scalesToTry = [1e-8,5e-8,1e-7,5e-7,1e-6,5e-6,1e-5,5e-5,1e-4,5e-4]\n",
    "sitesDiffByScale = []\n",
    "for scale in scalesToTry:\n",
    "    dSmoothedTest = dThresh.smooth(scale)\n",
    "    kdiffs = diem.count_site_differences(dSmoothedTest.DMBC,dThresh.DMBC)\n",
    "    sitesDiffByScale.append(kdiffs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "704c66ce-91c4-41ff-a4ca-6a3d278ea412",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,4))\n",
    "plt.plot(scalesToTry,sitesDiffByScale,marker='.')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff30b64e-aa9c-4df3-ac0a-2b898c2b1e2f",
   "metadata": {},
   "source": [
    "We see that by around scale = 5e-4, the single-site changes and short tracts have been smoothed over, and we choose this as our final smoothing scale.\n",
    "\n",
    "**note** smoothing returns a copy of the dataset with the same order (i.e. HIs are not recomputed and data is not resorted). That way, comparisons can be made easily before and after smoothing. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ee8f3c8-162d-418f-9147-ecc94dc3422f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dSmoothed = dThresh.smooth(5e-5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6f0d690-66b2-4771-ba6e-9e9a1c75b9d6",
   "metadata": {},
   "source": [
    "We can see how smoothing has affected our HIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6682fbe1-f65d-4e50-9a72-e6f74c9ea1f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,4))\n",
    "plt.plot(dThresh.HIs,marker='.',color='b')\n",
    "plt.plot(dSmoothed.computeHIs(),marker='.',color='r')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dc59681-2135-4bba-9d6d-373f45bb2bde",
   "metadata": {},
   "source": [
    " and we can plot the effect of smoothing on the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89e8f61d-9d23-4620-92e7-391d36158467",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(dSmoothed.DMBC[9])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "468f85ae-787e-43dd-a8e6-f68aa7ca9fb8",
   "metadata": {},
   "source": [
    "# Constructing Contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd6d0853-3563-4b28-bce0-394f0516cdfe",
   "metadata": {},
   "source": [
    "Downstream analyses of diem output may focus on the lengths of tracts of shared ancestry across the gene flow barrier.\n",
    "\n",
    "Here, we provide the function create_contig_matrix() that generates 'contigs' for each chromosome of each individual.\n",
    "\n",
    "a 'contig' is a summary of the data.  It is the sequence of contiguous intervals that form a givien chromosome \n",
    "\n",
    "Once we have created the contigs, we can export per-individual contig files in bed formatting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7806b16-d8f5-40eb-8de6-f5ea9d0c94a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dSmoothed.create_contig_matrix()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dfd71fd-edd5-43dc-bf26-9b560e47e4e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "contingOutputDirectory = './output/contigs/'\n",
    "\n",
    "diem.export_contigs_to_ind_bed_files(dSmoothed,contingOutputDirectory)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "192bb620-51f5-485a-a74b-19e63e1865ea",
   "metadata": {},
   "source": [
    "# Using Contigs & intervals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50367fc1-9d68-44af-90cf-7be4ed24c513",
   "metadata": {},
   "source": [
    "Within the diemtype objects, contigs are stored in the *contigMatrix*\n",
    "\n",
    "the *contigMatrix* is an array wit shape nChrs x nIndividuals. and we can use splicing to access specific contigs. \n",
    "\n",
    "For chromosome indexed 1 and individual indexed 8:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c19d201a-556a-4809-ae6d-8162473f83a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleContig = dSmoothed.contigMatrix[1,8]\n",
    "exampleContig.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25381047-eac6-4077-8d9a-36e8f576ca21",
   "metadata": {},
   "source": [
    "A contig stores the individual and chromosome names to which it belongs as well as a list of intervals and the number of intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29bd8323-1863-4327-9f39-94d9c027eb7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "?exampleContig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb733a6-55b4-4cbf-a2a2-c7c2b0639778",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleContig.num_intervals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95a7b300-f1a2-45e6-974d-6ca6cdf47e15",
   "metadata": {},
   "source": [
    "an **interval** is a contiguous segment of genome with the same ancestry.\n",
    "\n",
    "It stores the chromosome and individual to which it belongs. the indices and positions of the left- and right-most markers,\n",
    "and the state of the individual: 0, 1, 2, or 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3036113a-61d3-404b-9f56-5ce1680aeedb",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval = exampleContig.intervals[0]\n",
    "?exampleInterval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2fc74d-c511-40a2-899f-21e82770dd75",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b567b6a3-8d99-45c0-aaa2-33bd26c06278",
   "metadata": {},
   "source": [
    "We can get the interval's span easily:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b706378-6734-4bf7-9458-e1c2210cd663",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval.span()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53d0f834-fcbb-485d-adb3-48fed5a6594b",
   "metadata": {},
   "source": [
    "And we can get the intervals map span (on a unit scale) if we provide the length of the chromsome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2ebba57-cbed-430f-97dc-fafca76a2c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleInterval.mapSpan(dSmoothed.chrLengths[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea5e546-cf8e-4693-be87-e09446d265d2",
   "metadata": {},
   "source": [
    "Here is the information about all the intervals that form the genotype of our example contig:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63deb48d-aeb6-4396-b149-14888eddd053",
   "metadata": {},
   "outputs": [],
   "source": [
    "[x.__dict__ for x in exampleContig.intervals]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78ca23ff-fab9-4f84-bd3e-f2ae1b243703",
   "metadata": {},
   "source": [
    "We can plot the tract length distribution for all intervals with state '1'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38727955-362e-445a-a6b8-21fbf880d0e4",
   "metadata": {},
   "source": [
    "# Plotting tract length distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83af4ef1-26dc-4621-9c53-7509a188472e",
   "metadata": {},
   "outputs": [],
   "source": [
    "oneTracts = dSmoothed.get_intervals_of_state(1)\n",
    "oneLengths = [x.span() for x in oneTracts]\n",
    "plt.hist(oneLengths,color='m')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96f8aafe-c3ce-4993-8534-26dcebc6110f",
   "metadata": {},
   "source": [
    "We can plot the distribution of the different types of tracts together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4858008-3413-4ada-b2c3-a19e46ace485",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist([x.span() for x in dSmoothed.get_intervals_of_state(1)],bins = 50,color='m')\n",
    "plt.hist([x.span() for x in dSmoothed.get_intervals_of_state(2)],bins=50,color='y')\n",
    "plt.hist([x.span() for x in dSmoothed.get_intervals_of_state(3)],bins=50,color='teal')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ec03da8-d603-40d3-ae09-1fcf93ab244a",
   "metadata": {},
   "source": [
    "# Advanced tract length distribution plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6069df15-bfb6-41a2-9025-9dd15ee94957",
   "metadata": {},
   "source": [
    "We can also subset individuals and chromosomes when getting intervavls of those types."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0b0881f-e992-4775-a45a-743b2b9468d4",
   "metadata": {},
   "source": [
    "the barrier splits the population at the steepest change in HI. We can get the index which splits the barrier in two as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5a341fa-4e48-4d22-9ea4-c09428eebb2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "nIndsLeft = np.argmax(dSmoothed.HIs[1:] - dSmoothed.HIs[0:-1])+1\n",
    "print(nIndsLeft)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c3a482-3eaa-4952-93e6-9cde113fc347",
   "metadata": {},
   "outputs": [],
   "source": [
    "leftIndividuals = np.arange(0,nIndsLeft)\n",
    "rightIndividuals = np.arange(nIndsLeft,len(dSmoothed.indNames))\n",
    "print(leftIndividuals)\n",
    "print(rightIndividuals)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253f5779-5d3c-47a5-a126-5d2b8def768c",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "We can plot the spans of tracts of type '1' for individuals on the 'left' (blue) and 'right' (red) sides of the barrier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10668415-f855-45d7-8f55-171509722a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "state1SpansLeftSide = [x.span() for x in dSmoothed.get_intervals_of_state(1,individualSubset=leftIndividuals)]\n",
    "state1SpansRightSide = [x.span() for x in dSmoothed.get_intervals_of_state(1,individualSubset=rightIndividuals)]\n",
    "\n",
    "plt.hist(state1SpansLeftSide,bins = 100, color = 'b',alpha = 0.5)\n",
    "plt.hist(state1SpansRightSide, bins = 100, color = 'r',alpha = 0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cd1727-8aeb-4dee-a89a-d11c0632c220",
   "metadata": {},
   "source": [
    "Similarly, we can plot the spans of type '3' comparing the left and right sides of the barrier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cc95ba0-dd05-4b0a-ad63-f739243dc3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "state3SpansLeftSide = [x.span() for x in dSmoothed.get_intervals_of_state(3,individualSubset=leftIndividuals)]\n",
    "state3SpansRightSide = [x.span() for x in dSmoothed.get_intervals_of_state(3,individualSubset=rightIndividuals)]\n",
    "\n",
    "plt.hist(state3SpansRightSide, bins = 100, color = 'r',alpha = 0.5)\n",
    "plt.hist(state3SpansLeftSide,bins = 100, color = 'b',alpha = 0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3468c8ed-34c9-4abb-96d3-040bc8b53a6b",
   "metadata": {},
   "source": [
    "# Plotting paintings with marker spacing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fddec2df-157d-44bf-a5d8-8ea9a39ca9e7",
   "metadata": {},
   "source": [
    "We have provided a function which plots the paintings but accounts for the physical spacing between markers. Note that the resolution may be too poor on a dense dataset to see the difference.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff6a1f9-4af9-4c9b-a9ff-f7b4f42c84f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(dSmoothed.DMBC[9])\n",
    "diem.plot_painting_with_positions(dSmoothed.DMBC[9],dSmoothed.posByChr[9])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8718b1c4-73b9-4fc8-927d-733b84de1c1b",
   "metadata": {},
   "source": [
    "# Advanced Polarizing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50d895fd",
   "metadata": {},
   "source": [
    "## Site and Individual Masks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e299f9fb-15d4-48fc-9503-34fd21da545f",
   "metadata": {},
   "source": [
    "When running diem, we may want to mask some individual and sites.  Here, masked sites/individuals do not *influence* the barrier (i.e. they don't contribute to the likelihood function). However, they *are* polarized during the EM step along with other sites and individuals. \n",
    "\n",
    "We have included two example exclusion files under the user_files subdirectory.\n",
    "\n",
    "Here, we mask individuals that were sampled outside of france and spain.\n",
    "\n",
    "We also (arbitrarily) mask all of chromosome 2 for this example as well as two single sites on chromosome 3\n",
    "\n",
    "Although we have masked both sites and individuals in this example, it is not necessary to specify both. Any masks have been added to the raw data will be used in polarization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35002a37-ff64-432b-8bf7-723ef429b147",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #### add information about individuals that do not influence polarization. We say 'excluded' but really mean 'masked'\n",
    "sitesMaskedFile = './user_files/masked_regions.bed'\n",
    "individualsMaskedFile = './user_files/masked_individuals.txt'\n",
    "\n",
    "dRawWithMasks = dRaw.copy()\n",
    "\n",
    "dRawWithMasks.add_individual_exclusions(individualsMaskedFile)\n",
    "dRawWithMasks.add_site_exclusions(sitesMaskedFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e01c7640-53bb-41e2-b55d-9aa60d31687f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dPolWithMasks = dRawWithMasks.polarize()\n",
    "dPolWithMasks.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c5ca746-198a-4762-ae7b-96d0e64d3142",
   "metadata": {},
   "outputs": [],
   "source": [
    "diem.plot_painting(dPol.DMBC[9])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ac020e-e846-483d-a193-1dced0731d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5,4))\n",
    "plt.plot(dPolWithMasks.HIs,marker='.',color='b')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17ec84b9",
   "metadata": {},
   "source": [
    "## Specifying a pre-defined null polarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4399cbb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dRawPredefinedInitial = diem.read_diem_bed(bedInFile,metaCorrected)\n",
    "dRawPredefinedInitial.add_initial_polarity('./user_files/user_initial_pol.txt')\n",
    "\n",
    "dPolPredefinedInitial = dRawPredefinedInitial.polarize(ncores=4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
